### PR TITLE
add ssl support for kafka source.

### PR DESCRIPTION
**Pull request by morvencao**
ref: https://github.com/ansible/event-driven-ansible/issues/61

Signed-off-by: morvencao <lcao@redhat.com>

### CODE REVIEW

File: plugins/event_source/kafka.py [hunk 1]
```diff
@@ -42,6 +51,17 @@ async def main(queue: asyncio.Queue, args: Dict[str, Any]):
         max_poll_records=1,
         auto_offset_reset=offset,
     )
+
+    if not cafile or not certfile or not keyfile or not password:
```
Comment (bzwei) [inline plugins/event_source/kafka.py:55]: What is the minimal requirement to run in SSL mode? Should the `if` condition be
```
if cafile or certfile or keyfile or password:
```

Comment (morvencao) [reply plugins/event_source/kafka.py]: right, `cafile` is required for SSL mode, while `certfile`, `keyfile` and `password` are optional.

Comment (bzwei) [reply plugins/event_source/kafka.py]: if `catfile` is required but others are optional, then the condition should be:
```
   if catfile:
```

Comment (morvencao) [reply plugins/event_source/kafka.py]: done.

File: plugins/event_source/kafka.py [hunk 2]
```diff
@@ -42,6 +51,17 @@ async def main(queue: asyncio.Queue, args: Dict[str, Any]):
         max_poll_records=1,
         auto_offset_reset=offset,
     )
+
+    if cafile:
+        context = create_ssl_context(
+            cafile=str(cafile),
+            certfile=str(certfile),
```
[the remark this answers is not in the document]
Comment (morvencao) [reply plugins/event_source/kafka.py]: thanks, I'm not an expert of python.
anyway, applied your comments.

**Review by bzwei — APPROVED** [review]

**Comment by ttuffin** [comment]
Hi @morvencao, please resolve the following linting issues:
```
ERROR: plugins/event_source/kafka.py:10:161: E501: line too long (173 > 160 characters)
ERROR: plugins/event_source/kafka.py:62:41: E225: missing whitespace around operator
ERROR: plugins/event_source/kafka.py:63:35: E225: missing whitespace around operator
```

**Comment by morvencao** [comment]
@ttuffin thanks for the reminder, hopefully now everything is fine.

**Comment by ttuffin** [comment]
Hi @morvencao, I am unable to get this implementation working against my local Kafka instance using SSL. Does it work for you?

**Comment by ttuffin** [comment]
I think there may be in issue with trying to assign the SSL context and protocol after `AIOKafkaConsumer` has already been instantiated. I believe those two params should be set when the consumer object is created. The following worked for me:
```
diff --git a/plugins/event_source/kafka.py b/plugins/event_source/kafka.py
index dd8c939..e5f9ba2 100644
--- a/plugins/event_source/kafka.py
+++ b/plugins/event_source/kafka.py
@@ -37,12 +37,22 @@ async def main(queue: asyncio.Queue, args: Dict[str, Any]):
     certfile = args.get("certfile")
     keyfile = args.get("keyfile")
     password = args.get("password")
+    check_hostname = args.get("check_hostname", True)
     group_id = args.get("group_id", None)
     offset = args.get("offset", "latest")

     if offset not in ("latest", "earliest"):
         raise Exception(f"Invalid offset option: {offset}")

+    if cafile:
+        context = create_ssl_context(
+            cafile=cafile,
+            certfile=certfile,
+            keyfile=keyfile,
+            password=password,
+        )
+        context.check_hostname=check_hostname
+
     kafka_consumer = AIOKafkaConsumer(
         topic,
         bootstrap_servers="{0}:{1}".format(host, port),
@@ -50,18 +60,10 @@ async def main(queue: asyncio.Queue, args: Dict[str, Any]):
         enable_auto_commit=True,
         max_poll_records=1,
         auto_offset_reset=offset,
+        security_protocol="SSL" if cafile else "PLAINTEXT",
+        ssl_context=context if cafile else None,
     )

-    if cafile:
-        context = create_ssl_context(
-            cafile=cafile,
-            certfile=certfile,
-            keyfile=keyfile,
-            password=password,
-        )
-        kafka_consumer.security_protocol="SSL"
-        kafka_consumer.ssl_context=context
-
     await kafka_consumer.start()
```

I also added another arg (check_hostname) to allow for disabling SSL hostname verification.

@morvencao @bzwei, what do you think?

**Comment by morvencao** [comment]
thank you @ttuffin 
Applied your changes.

File: plugins/event_source/kafka.py [hunk 3]
```diff
@@ -6,6 +6,12 @@
 Arguments:
     host:      The host where the kafka topic is hosted
     port:      The port where the kafka server is listening
+    cafile     The optional certificate authority file path containing certificates used to
+               sign kafka broker certificates
+    certfile   The optional client certificate file path containing the client certificate,
+               as well as CA certificates needed to establish the certificate's authenticity
+    keyfile    The optional client key file path containing the client private key
+    password   The optional password to be used when loading the certificate chain
```
Comment (ttuffin) [inline plugins/event_source/kafka.py:14]:
```suggestion
    password   The optional password to be used when loading the certificate chain
    check_hostname   Enable SSL hostname verification. [True (default), False]
```

**Review by ttuffin — CHANGES_REQUESTED** [review]
Need to add `check_hostname` to docstring.

File: plugins/event_source/kafka.py [hunk 4]
```diff
@@ -28,20 +36,37 @@ async def main(queue: asyncio.Queue, args: Dict[str, Any]):
...
         raise Exception(f"Invalid offset option: {offset}")
 
+    if cafile:
+        context = create_ssl_context(
+            cafile=cafile,
+            certfile=certfile,
+            keyfile=keyfile,
+            password=password,
+        )
+        context.check_hostname=check_hostname
```
Comment (ttuffin) [inline plugins/event_source/kafka.py:57]: Apologies @morvencao, we need whitespace around the operator to fix the linting issue.

**Review by ttuffin — CHANGES_REQUESTED** [review]

File: plugins/event_source/kafka.py (hunk 4 above)
Comment (morvencao) [reply plugins/event_source/kafka.py]: done

**Review by ttuffin — DISMISSED** [review]

**Review by bzwei — DISMISSED** [review]

**Comment by morvencao** [comment]
rebased the change set to fix the integration test.

**Review by bzwei — APPROVED** [review]

**Comment by bzwei** [comment]
This PR can be merged because the integration test failure is not caused by it.

**Review by ttuffin — APPROVED** [review]
Thanks @morvencao

**Pull request merged**